MailManager/src/test/java/inc/softserve/EnvelopeServiceTest.java

```diff
@@ -0,0 +1,76 @@
+package inc.softserve;
+
+import inc.softserve.common.EnvelopeTools;
+import inc.softserve.dao.EnvelopeDao;
+import inc.softserve.dao.UserDao;
+import inc.softserve.services.EnvelopeService;
+import inc.softserve.services.EnvelopeServiceImpl;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class EnvelopeServiceTest extends Assert {
+
+    private EnvelopeService service;
+    private UserDao usrDao;
+    private EnvelopeDao envDao;
+    private EnvelopeTools envTools;
+
+    @BeforeClass
+    public void setUpDataSource() throws Exception {
+        usrDao = mock(UserDao.class);
+        envDao = mock(EnvelopeDao.class);
+        envTools = mock(EnvelopeTools.class);
+
+        service = new EnvelopeServiceImpl(envDao, usrDao, envTools);
+    }
+
+    @DataProvider(name = "saveOrUpdateData")
+    public Object[][] saveOrUpdateData() {
+        return new Object[][]{
+                {EnvelopeState.SENT},
+                {EnvelopeState.IN_PROGRESS},
+                {EnvelopeState.ERROR},
+        };
+    }
+
+    @Test(dataProvider = "saveOrUpdateData")
+    public void testSaveOrUpdate(EnvelopeState expected) throws Exception {
+        User user = new User("testUser", "a", "b");
+
+        when(usrDao.getUserByUsername("testUser")).thenReturn(user);
+        when(envDao.saveOrUpdate(any(Envelope.class))).thenAnswer(args -> args.getArguments()[0]); //returns same argument passed to saveOrUpdate() method
```

Comment:
It might be better to say `returns object` rather then `argument`. Because the fact that it returns argument is obvious from the code itself. While it might have sense to highlight the fact that mock DAO method returns the object itself upon call to `saveOrUpdate`.